File: go/deadline.go
package dialcache

import (
	"sync"
	"time"
)

type completion[T any] struct {
	value T
	err   error
}
type pending[T any] struct {
	done   chan struct{}
	result completion[T]
}

func callSafely[T any](f func() (T, error)) (value T, err error) {
	defer func() {
		if p := recover(); p != nil {
			err = &CallbackPanicError{Value: p}
		}
	}()
	return f()
}

func startPending[T any](f func() (T, error)) *pending[T] {
	p := &pending[T]{done: make(chan struct{})}
	go func() { p.result.value, p.result.err = callSafely(f); close(p.done) }()
	return p
}

func after(clock Clock, ms int64, f func()) Timer {
	if ms < 0 {
		ms = 0
	}
	if timers, ok := clock.(TimerClock); ok {
		return timers.AfterFunc(ms, f)
	}
	return time.AfterFunc(time.Duration(ms)*time.Millisecond, f)
}

// awaitDeadline accepts only results observed strictly before the deadline.
// Raw work keeps ownership of its resources after the caller stops waiting.
func awaitDeadline[T any](clock Clock, p *pending[T], started, budget int64, timeout func() error, onTimeout func()) (T, error) {
	if budget < 0 {
		<-p.done
		return p.result.value, p.result.err
	}
	var once sync.Once
	expired := make(chan struct{})
	timer := after(clock, budget-(clock.ElapsedMS()-started), func() { once.Do(func() { close(expired) }) })
	defer timer.Stop()
	select {
	case <-p.done:
		if clock.ElapsedMS()-started < budget {
			return p.result.value, p.result.err
		}
	case <-expired:
	}
	if onTimeout != nil {
		onTimeout()
	}
	var zero T
	return zero, timeout()
}

func deferWork(clock Clock, f func()) {
	if executor, ok := clock.(DeferredExecutor); ok {
		executor.Defer(f)
		return
	}
	go f()
}

File: rust/src/deadline.rs
//! Monotonic deadlines over controlled clocks.

use std::time::Duration;

use futures::future::{select, Either};

use crate::clock::Clock;
use crate::flight::Settled;

/// Elapsed time since `start`, never negative.
pub(crate) fn since(clock: &dyn Clock, start: Duration) -> Duration {
    clock.elapsed().saturating_sub(start)
}

pub(crate) fn seconds_since(clock: &dyn Clock, start: Duration) -> f64 {
    since(clock, start).as_secs_f64()
}

/// Wait for `pending`, accepting only a result observed strictly before the
/// deadline `started + budget_ms`. Raw work keeps its resources after the
/// caller stops waiting. `None` budget waits without bound.
pub(crate) async fn await_deadline<T: Clone>(
    clock: &dyn Clock,
    pending: &Settled<T>,
    started: Duration,
    budget_ms: Option<u64>,
    on_timeout: impl FnOnce() -> T,
) -> T {
    let Some(budget_ms) = budget_ms else {
        return pending.wait().await;
    };
    let budget = Duration::from_millis(budget_ms);
    loop {
        if let Some(value) = pending.peek() {
            if since(clock, started) < budget {
                return value;
            }
            break;
        }
        let remaining = budget.saturating_sub(since(clock, started));
        let timer = clock.sleep(remaining);
        match select(pending.wait(), timer).await {
            Either::Left((value, _timer)) => {
                if since(clock, started) < budget {
                    return value;
                }
                break;
            }
            Either::Right(((), _wait)) => {
                // Timer precision and delivery do not define the semantic boundary.
                if since(clock, started) < budget {
                    continue;
                }
                break;
            }
        }
    }
    on_timeout()
}
